### ISSUE TITLE
Full compatibility with Segment Anything

### ISSUE THREAD

**Issue by astnmsn** [tracel-ai/burn#1544]
I have not found any equivalent request

### Feature description

Addition of support for necessary operators to utilize the `vit_b` SAM Model which can be found [here](https://github.com/facebookresearch/segment-anything?tab=readme-ov-file#model-checkpoints): 

I have inspected the model using Netron and compared the nodes to the support list [here](https://github.com/tracel-ai/burn/blob/main/crates/burn-import/SUPPORTED-ONNX-OPS.md)

Below is the list of operators in the network that are partially or fully missing from the support list
Operator (missing support)

- Mul (Import)
- MatMul (Import)
- Sin (Import)
- Shape (Import)
- Expand (Full)
- Not (Import)
- ReduceMean (Import)
- ConstantOfShape (Full)
- Where (Import)
- Slice (Import)
- OneHot (Import)
- Tile (Full)
- LayerNormalization (Import)
- Gemm (Full)
- ReduceMax (Import)
- Floor (Full)

### Feature motivation

I am currently trying to load and run an onnx model [segment-anything](https://github.com/facebookresearch/segment-anything) inside an image editing app in an effort to provide a masking experience similar to the demo available [here](https://segment-anything.com/demo). I am integrating it into an existing rust codebase using wgpu that compiles to wasm and runs in the browser.

Before I can select `burn` as the ml library to support this workflow, I need to be sure that it supports the operators specified in the model.

**Comment by antimora** [comment]
Thanks for filing this. This is helpful as we prioritize ONNX ops. If you have a direct link to the ONNX file, can you also link this?

**Comment by antimora** [comment]
Updating Expand to Import, since we just added this op. I need to update the docs.

**Comment by antimora** [comment]
Submitted a PR to fix the supported OPs document: https://github.com/tracel-ai/burn/pull/1547

**Comment by astnmsn** [comment]
This is the model [download link](https://dl.fbaipublicfiles.com/segment_anything/sam_vit_b_01ec64.pth) provided by the SAM repo

**Comment by antimora** [comment]
I think it might be faster to implement the model manually in Burn and load the pth weights file, which we now support.

You can check out an existing model to see how it's done: https://github.com/tracel-ai/models/tree/main/resnet-burn

We also have a YOLOX object detection PR in the works: https://github.com/tracel-ai/models/pull/24

@laggui has written a great tutorial on this subject: https://dev.to/laggui/transitioning-from-pytorch-to-burn-45m

Recently, we made tons of enhancements to the PyTorchFileRecorder: https://discord.com/channels/1038839012602941528/1144670451763785769/1216788417984335872

<img width="703" alt="image" src="https://github.com/tracel-ai/burn/assets/939125/fb6461e8-f9cf-490c-9f1e-20d1e7dfdb59">



@laggui, @nathanielsimard, @ashdtu, would this be worth implementing ourselves? Should we move this ticket to the models repo?

**Comment by laggui** [comment]
The community is always one step ahead 😄

We've actually discussed adding SAM to our models and this was in the plans following the release.

We still haven't decided whether we want to reimplement it and use the PyTorch file recorder to import the weights or use the ONNX import.

**Comment by antimora** [comment]
@laggui, if we decide to work on this, I am more inclined to adding ONNX OPs. It will be biggest bang for the buck instead of spending time to come up with the model by hand (although I am not sure how complex it is).

**Comment by laggui** [comment]
Btw, not sure if anyone has delved into the SAM code for ONNX export but it doesn't include all the operations to actually run the model for an input image. The encoder part is totally left out of the ONNX export and the exported ONNX model expects image embeddings as input.

In their example they still use their pytorch implementation to provide the embeddings to the ONNX runtime.

So even if we support the missing operations in this issue, SAM support will still not be complete. Is this what you expected @astnmsn?

**Comment by astnmsn** [comment]
@laggui Thanks for asking, and yes that is expected. We plan to run the first half of the model to generate the embeddings on the backend using pytorch. Only the second half, which produces the masks from the embeddings and the cursor/click positions, will be run on the client

**Comment by antimora** [comment]
Regarding Tile Op. We need to rename our current repeat op to repeat_dim and implement a proper repeat for all dimensions at once.

**Comment by antimora** [comment]
Resolving this ticket will resolve https://github.com/tracel-ai/burn/issues/1560 as well.

**Comment by laggui** [comment]
Current state of required ops based on the latest PRs:

| op_type            | Burn | Import |
| :----------------- | :--: | :----: |
| Add                | ✅  | ✅     |
| Cast               | ✅  | ✅     |
| Concat             | ✅  | ✅     |
| Constant           | ✅  | ✅     |
| ConstantOfShape    | ❌  | ❌     |
| Conv               | ✅  | ✅     |
| ConvTranspose      | ✅  | ✅     |
| Cos                | ✅  | ✅     |
| Div                | ✅  | ✅     |
| Equal              | ✅  | ✅     |
| Erf                | ✅  | ✅     |
| Expand             | ✅  | ❌     |
| Floor              | ❌  | ❌     |
| Gather             | ✅  | ✅     |
| Gemm               | ❌  | ❌     |
| LayerNormalization | ✅  | ✔️     |
| MatMul             | ✅  | ✔️     |
| Mul                | ✅  | ✅     |
| Not                | ✅  | ✔️     |
| OneHot             | ✅  | ❌     |
| Pow                | ✅  | ✅     |
| Reciprocal         | ✅  | ✅     |
| ReduceMax          | ✅  | ✔️     |
| ReduceMean         | ✅  | ✔️     |
| Relu               | ✅  | ✅     |
| Reshape            | ✅  | ✅     |
| Resize             | ✅  | ❌     |
| Shape              | ✅  | ✔️     |
| Sin                | ✅  | ✔️     |
| Slice              | ✅  | ❌     |
| Softmax            | ✅  | ✅     |
| Sqrt               | ✅  | ✅     |
| Sub                | ✅  | ✅     |
| Tile               | ✅  | ❌     |
| Transpose          | ✅  | ✅     |
| Unsqueeze          | ✅  | ✅     |
| Where              | ✅  | ✔️     |

**Comment by antimora** [comment]
Can confirm this is still an issue:

```
}), value: None, passed: false } for const
DEBUG onnx_ir::from_onnx: Input constant49_out1 matched constant node constant49
DEBUG onnx_ir::from_onnx: Lifted and removed constant node constant49 for reshape2 (single use)
DEBUG onnx_ir::from_onnx: Skipping constant creation for reshape2 input 1 (will be lifted)
DEBUG onnx_ir::rank_inference: Inferring rank for node: reshape2
ERROR burn_import::logger: PANIC => panicked at crates/onnx-ir/src/node/reshape.rs:55:14:
Reshape: invalid input type - expected Tensor or Shape, got Scalar(Int64)

thread 'main' panicked at crates/onnx-ir/src/node/reshape.rs:55:14:
Reshape: invalid input type - expected Tensor or Shape, got Scalar(Int64)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[burn-import]%
```

**Comment by HiDolen** [comment]
> We merged a fix for Reshape of scalar today: [#4119](https://github.com/tracel-ai/burn/pull/4119)
> 
> Can you check if this works for you?

Thank you for such a prompt reply!

I just pulled the latest code from the main branch using `burn = { git = "https://github.com/tracel-ai/burn.git ", branch = "main", features = ["ndarray"] }`, but I’m still getting exactly the same error.

My ONNX model is unusually large for some reason, and I’m currently looking into the cause. Once I figure it out, I might be able to upload an example model.

**Comment by akshitgaur2005** [comment]
I was having this problem for SAM2, but upgrading to the git branch changed the error to this-

```
  DEBUG onnx_ir::phases::node_conversion: Preserving outer-scope value_store for 'constant141_out1' (Constant)
   WARN onnx_ir::graph_state: Input  not found, should only happen when peeking
   WARN onnx_ir::graph_state: Input  not found, should only happen when peeking
  DEBUG onnx_ir::pipeline:  PHASE 3: Type Inference 
  ERROR burn_import::logger: PANIC => panicked at /home/akshit/.cargo/git/checkouts/burn-6c277d792b0d5d7a/f1e90cf/crates/burn-import/src/onnx/model_gen.rs:296:33:
  Failed to parse ONNX file 'src/models/2.1/tiny/decoder_opset16.onnx': Type inference failed: TypeMismatch { expected: "Tensor", actual: "Shape(1)" }

  --- stderr

  thread 'main' (70) panicked at /home/akshit/.cargo/git/checkouts/burn-6c277d792b0d5d7a/f1e90cf/crates/burn-import/src/onnx/model_gen.rs:296:33:
  Failed to parse ONNX file 'src/models/2.1/tiny/decoder_opset16.onnx': Type inference failed: TypeMismatch { expected: "Tensor", actual: "Shape(1)" }
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
```

The model I am trying out is [sam2.1_hiera_tiny_decoder.onnx](https://huggingface.co/mabote-itumeleng/ONNX-SAM2-Segment-Anything/blob/main/sam2.1_hiera_tiny_decoder.onnx)
